### PR TITLE
Rewriting bash in Go in multiple modules

### PR DESCRIPTION
**Pull request by zuzzas**
## Description

Rewrote multiple bash scripts in container image in Go.

## Why do we need it, and what problem does it solve?

We'd like to have self-contained initialization and helper binaries.

## What is the expected result?

Some modules' images contain only compiled binaries.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: chore
summary: Rewrote multiple helper scripts into compiled binaries written in Go
impact_level: low
```

### CODE REVIEW

File: modules/021-kube-proxy/images/init-container/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,12 @@
-ARG BASE_ALPINE
-FROM $BASE_ALPINE
-RUN apk add --no-cache bash sed curl jq
-COPY api-proxy-check.sh nodeport-bind-address.sh /
+ARG BASE_GOLANG_19_ALPINE
+FROM $BASE_GOLANG_19_ALPINE as artifact
+
+WORKDIR /src/
+COPY main.go go.mod go.sum /src/
+RUN apk add --no-cache git && \
```
Comment (RomanenkoDenys) [inline modules/021-kube-proxy/images/init-container/Dockerfile:6]: why we need git here?

File: modules/021-kube-proxy/images/init-container/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,12 @@
...
+    GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -w" -o init main.go
+
+FROM scratch
```
Comment (RomanenkoDenys) [inline modules/021-kube-proxy/images/init-container/Dockerfile:9]: we have base image for this - BASE_SCRATCH

File: modules/021-kube-proxy/images/init-container/main.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+)
+
+func main() {
+	podSubnet, ok := os.LookupEnv("POD_SUBNET")
+	if !ok {
+		log.Fatal("POD_SUBNET env not provided")
+	}
+	controlPlaneAddress, ok := os.LookupEnv("CONTROL_PLANE_ADDRESS")
+	if !ok {
+		log.Fatal("CONTROL_PLANE_ADDRESS not provided")
```
Comment (RomanenkoDenys) [inline modules/021-kube-proxy/images/init-container/main.go:37]:
```suggestion
		log.Fatal("CONTROL_PLANE_ADDRESS env not provided")
```

File: modules/035-cni-flannel/images/flanneld/Dockerfile [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
 FROM $BASE_ALPINE
 
 COPY --from=base /opt/bin/flanneld /opt/bin/
 COPY --from=base /opt/bin/mk-docker-opts.sh /opt/bin/
 
-COPY entrypoint.sh /
+COPY --from=artifact /src/entrypoint /
+
 COPY iptables-wrapper-installer.sh /
 RUN apk add --no-cache curl jq
```
Comment (RomanenkoDenys) [inline modules/035-cni-flannel/images/flanneld/Dockerfile:21]: why we need curl and jq ?

File: modules/035-cni-flannel/images/flanneld/main.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+		err := deleteLinksByPrefix("flannel")
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (RomanenkoDenys) [inline modules/035-cni-flannel/images/flanneld/main.go:40]: i think this if should be at the top of main func

File: modules/042-kube-dns/images/resolv-watcher/Dockerfile [hunk 6]
```diff
@@ -1,5 +1,11 @@
...
-ENTRYPOINT /resolv-conf-check
+ARG BASE_GOLANG_19_ALPINE
+
+FROM $BASE_GOLANG_19_ALPINE as artifact
+WORKDIR /src
+COPY . .
+RUN CGO_ENABLED=0 go build -ldflags="-s -w" -o resolv-watcher main.go
+
+
+FROM scratch
```
Comment (RomanenkoDenys) [inline modules/042-kube-dns/images/resolv-watcher/Dockerfile:9]: BASE_SCRATCH

File: modules/042-kube-dns/images/sts-pods-hosts-appender-webhook/Dockerfile [hunk 7]
```diff
@@ -1,11 +1,9 @@
...
-FROM $BASE_GOLANG_16_ALPINE as artifact
+ARG BASE_GOLANG_19_ALPINE
+FROM $BASE_GOLANG_19_ALPINE as artifact
 WORKDIR /src/
 COPY main.go go.mod go.sum /src/
 RUN apk add --no-cache git && \
     GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -w" -o sts-pods-hosts-appender-webhook main.go
 
-FROM $BASE_ALPINE
+FROM scratch
```
Comment (RomanenkoDenys) [inline modules/042-kube-dns/images/sts-pods-hosts-appender-webhook/Dockerfile:8]: BASE_SCRATCH

File: modules/340-monitoring-kubernetes/images/kube-state-metrics/Dockerfile [hunk 8]
```diff
@@ -9,6 +9,6 @@ RUN mkdir -p /src/kube-state-metrics && \
 WORKDIR /src/kube-state-metrics
 RUN make build-local
 
-FROM $BASE_ALPINE
+FROM scratch
```
Comment (RomanenkoDenys) [inline modules/340-monitoring-kubernetes/images/kube-state-metrics/Dockerfile:12]: BASE_SCRATCH

File: modules/402-ingress-nginx/images/protobuf-exporter/Dockerfile [hunk 9]
```diff
@@ -4,9 +4,9 @@ FROM $BASE_GOLANG_16_ALPINE as artifact
 WORKDIR /src/
 COPY / /src/
 RUN apk add --no-cache git && \
-    GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -w" -o ./protobuf_exporter ./main.go
+    GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags='-extldflags "-static" -s -w' -o ./protobuf_exporter ./main.go
 
-FROM $BASE_ALPINE
+FROM scratch
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/protobuf-exporter/Dockerfile:9]: BASE_SCRATCH

File: modules/402-ingress-nginx/images/proxy-failover/main.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	if len(controllerName) == 0 {
+		log.Fatal("CONTROLLER_NAME env is empty")
+	}
+
+	nginxConfTemplate, err := os.ReadFile("/etc/nginx/nginx.conf.tpl")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	t := template.Must(template.New("nginxTemplate").Parse(string(nginxConfTemplate)))
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/proxy-failover/main.go:22]: You can use os.ExpandEnv here

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
some nits

File: modules/402-ingress-nginx/images/proxy-failover/main.go (hunk 10 above)
Comment (zuzzas) [reply modules/402-ingress-nginx/images/proxy-failover/main.go]: Done.

File: modules/035-cni-flannel/images/flanneld/Dockerfile (hunk 4 above)
Comment (zuzzas) [reply modules/035-cni-flannel/images/flanneld/Dockerfile]: Fixed.

File: modules/035-cni-flannel/images/flanneld/main.go (hunk 5 above)
Comment (zuzzas) [reply modules/035-cni-flannel/images/flanneld/main.go]: Done.

File: modules/021-kube-proxy/images/init-container/main.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+				},
+			},
+		},
+		CurrentContext: "default",
+	}
+
+	kubeProxyConfig := &v1alpha1.KubeProxyConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "KubeProxyConfiguration",
+			APIVersion: "kubeproxy.config.k8s.io/v1alpha1",
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply modules/021-kube-proxy/images/init-container/main.go]: Fixed.

File: modules/021-kube-proxy/images/init-container/main.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+	err = os.WriteFile(kubeConfigPath, kubeConfigBytes, 0666)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = os.WriteFile(kubeProxyConfigPath, kubeProxyConfigBytes, 0666)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
```
Comment (RomanenkoDenys) [inline modules/021-kube-proxy/images/init-container/main.go:125]:
```suggestion
	err = os.WriteFile(kubeConfigPath, kubeConfigBytes, 0644)
	if err != nil {
		log.Fatal(err)
	}
	err = os.WriteFile(kubeProxyConfigPath, kubeProxyConfigBytes, 0644)
	if err != nil {
		log.Fatal(err)
	}
}
```

File: modules/402-ingress-nginx/images/proxy-failover/rootfs/etc/nginx/nginx.conf.tpl [hunk 13]
```diff
@@ -38,7 +38,7 @@ stream {
   proxy_protocol on;
 
   upstream http {
-    server controller-${CONTROLLER_NAME}-failover:80 max_fails=0;
+    server controller-${CONTROLLER_NAME} -failover:80 max_fails=0;
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/proxy-failover/rootfs/etc/nginx/nginx.conf.tpl:41]:
```suggestion
    server controller-${CONTROLLER_NAME}-failover:80 max_fails=0;
```

**Review by RomanenkoDenys — COMMENTED** [review]
some nits

File: modules/035-cni-flannel/images/flanneld/main.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+		}
+	}
+
+	var allIPs []string
+	internalIPs, externalIPs, err := getInternalAndExternalIPs(hostname)
+	if err != nil {
+		log.Fatal(err)
+	}
+	allIPs = append(allIPs, internalIPs...)
+	allIPs = append(allIPs, externalIPs...)
```
Comment (RomanenkoDenys) [inline modules/035-cni-flannel/images/flanneld/main.go:64]:
```suggestion
	allIPs = append(allIPs, internalIPs..., externalIPs...)
```

File: modules/042-kube-dns/images/resolv-watcher/main.go [hunk 15]
```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2022 Flant JSC
```
Comment (RomanenkoDenys) [inline modules/042-kube-dns/images/resolv-watcher/main.go:2]:
```suggestion
Copyright 2023 Flant JSC
```

File: modules/042-kube-dns/images/sts-pods-hosts-appender-init-container/main.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2022 Flant JSC
```
Comment (RomanenkoDenys) [inline modules/042-kube-dns/images/sts-pods-hosts-appender-init-container/main.go:2]:
```suggestion
Copyright 2023 Flant JSC
```

File: modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main.go [hunk 17]
```diff
@@ -0,0 +1,378 @@
+/*
+Copyright 2022 Flant JSC
```
Comment (RomanenkoDenys) [inline modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main.go:2]:
```suggestion
Copyright 2023 Flant JSC
```

File: modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main_test.go [hunk 18]
```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2022 Flant JSC
```
Comment (RomanenkoDenys) [inline modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main_test.go:2]:
```suggestion
Copyright 2023 Flant JSC
```

File: modules/402-ingress-nginx/images/proxy-failover-iptables/main.go [hunk 19]
```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2022 Flant JSC
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/proxy-failover-iptables/main.go:2]:
```suggestion
Copyright 2023 Flant JSC
```

File: modules/402-ingress-nginx/images/proxy-failover/main.go [hunk 20]
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2022 Flant JSC
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/proxy-failover/main.go:2]:
```suggestion
Copyright 2023 Flant JSC
```

**Review by RomanenkoDenys — COMMENTED** [review]
LGTM

File: modules/402-ingress-nginx/images/proxy-failover/main.go (hunk 20 above)
Comment (zuzzas) [reply modules/402-ingress-nginx/images/proxy-failover/main.go]: Fixed.

File: modules/402-ingress-nginx/images/proxy-failover-iptables/main.go (hunk 19 above)
Comment (zuzzas) [reply modules/402-ingress-nginx/images/proxy-failover-iptables/main.go]: Fixed.

File: modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main_test.go (hunk 18 above)
Comment (zuzzas) [reply modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main_test.go]: Fixed.

File: modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main.go (hunk 17 above)
Comment (zuzzas) [reply modules/340-monitoring-kubernetes/images/kubelet-eviction-thresholds-exporter/main.go]: Fixed.

File: modules/042-kube-dns/images/sts-pods-hosts-appender-init-container/main.go (hunk 16 above)
Comment (zuzzas) [reply modules/042-kube-dns/images/sts-pods-hosts-appender-init-container/main.go]: Fixed.

File: modules/042-kube-dns/images/resolv-watcher/main.go (hunk 15 above)
Comment (zuzzas) [reply modules/042-kube-dns/images/resolv-watcher/main.go]: Fixed.

**Review by RomanenkoDenys — APPROVED** [review]
lgtm

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/402-ingress-nginx/images/proxy-failover-iptables/main.go [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+			}
+		}
+	}
+}
+
+func loop(iptablesMgr *iptables.IPTables) error {
+	resp, err := http.Get("http://127.0.0.1:10254/healthz")
+	defer func(Body io.ReadCloser) {
+		_ = Body.Close()
+	}(resp.Body)
```
Comment (yalosev) [inline modules/402-ingress-nginx/images/proxy-failover-iptables/main.go:139]:
```suggestion
	defer resp.Body.Close()
```

**Review by yalosev — DISMISSED** [review]

File: modules/402-ingress-nginx/images/proxy-failover-iptables/main.go (hunk 21 above)
Comment (RomanenkoDenys) [reply modules/402-ingress-nginx/images/proxy-failover-iptables/main.go:139]: it is go lint fix to avoid unprocessed error warning

**Review by yalosev — DISMISSED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/static'...

**Comment by zuzzas** [comment]
/e2e/abort static;Static;containerd;1.23 4741376518 4741376518-1-con-1-23 /sys/deckhouse-oss/install:pr4235 astra@192.168.199.48

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/static'...

**Review by yalosev — APPROVED** [review]

**Comment by zuzzas** [comment]
/e2e/abort static;Static;containerd;1.23 4752007649 4752007649-1-con-1-23 /sys/deckhouse-oss/install:pr4235 astra@192.168.199.38

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've recognized your '/e2e/abort' command and started the workflow...

**Pull request merged**